// .ignored/@ts-safeql/eslint-plugin/lib/utils/fp-ts.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.J = exports.TE = exports.O = exports.pipe = exports.identity = exports.flow = exports.E = exports.json = void 0;
var fp_ts_1 = require("fp-ts");
Object.defineProperty(exports, "json", { enumerable: true, get: function () { return fp_ts_1.json; } });
exports.E = __importStar(require("fp-ts/Either"));
var function_1 = require("fp-ts/function");
Object.defineProperty(exports, "flow", { enumerable: true, get: function () { return function_1.flow; } });
Object.defineProperty(exports, "identity", { enumerable: true, get: function () { return function_1.identity; } });
Object.defineProperty(exports, "pipe", { enumerable: true, get: function () { return function_1.pipe; } });
exports.O = __importStar(require("fp-ts/Option"));
exports.TE = __importStar(require("fp-ts/TaskEither"));
exports.J = __importStar(require("fp-ts/Json"));
